src/LambdaBTree.h

```diff
@@ -0,0 +1,654 @@
+        }
+
+        while (true) {
+            // handle inner nodes
+            if (cur->inner) {
+                auto a = &(cur->keys[0]);
+                auto b = &(cur->keys[cur->numElements]);
+
+                auto pos = this->search.lower_bound(k, a, b, this->weak_comp);
+                auto idx = pos - a;
+
+                // early exit for sets
+                if (isSet && pos != b && this->weak_equal(*pos, k)) {
+                    // update provenance information
+                    if (typeid(Comparator) != typeid(WeakComparator) && this->less(k, *pos)) {
+                        this->update(*pos, k);
+                        return (*pos).second;
+                    }
+
+                    return (*pos).second;
+                }
+
+                cur = cur->getChild(idx);
+                continue;
+            }
+
+            // the rest is for leaf nodes
+            assert(!cur->inner);
+
+            // -- insert node in leaf node --
+
+            auto a = &(cur->keys[0]);
+            auto b = &(cur->keys[cur->numElements]);
+
+            auto pos = this->search.upper_bound(k, a, b, this->weak_comp);
+            auto idx = pos - a;
+
+            // early exit for sets
+            if (isSet && pos != a && this->weak_equal(*(pos - 1), k)) {
+                // update provenance information
+                if (typeid(Comparator) != typeid(WeakComparator) && this->less(k, *(pos - 1))) {
+                    this->update(*(pos - 1), k);
+                    return (*(pos - 1)).second;
+                }
+
+                return (*(pos - 1)).second;
+            }
+
+            if (cur->numElements >= parenttype::node::maxKeys) {
+                // split this node
+                idx -= cur->rebalance_or_split(
+                        const_cast<typename parenttype::node**>(&this->root), this->root_lock, idx);
+
+                // insert element in right fragment
+                if (((typename parenttype::size_type)idx) > cur->numElements) {
+                    idx -= cur->numElements + 1;
+                    cur = cur->parent->getChild(cur->position + 1);
+                }
+            }
+
+            // ok - no split necessary
+            assert(cur->numElements < parenttype::node::maxKeys && "Split required!");
+
+            // move keys
+            for (int j = cur->numElements; j > idx; --j) {
+                cur->keys[j] = cur->keys[j - 1];
+            }
+
+            // call the functor as we've successfully inserted
+            typename Functor::result_type res = f(k);
+            // insert new element
+            cur->keys[idx] = k;
+            cur->numElements++;
+
+            // remember last insertion position
+            hints.last_insert.access(cur);
+            return res;
+        }
+#endif
+    }
+
+    /**
+     * Inserts the given range of elements into this tree.
+     */
+    template <typename Iter>
+    void insert(const Iter& a, const Iter& b) {
```

Comment:
Do you need to reimplement all the inherited methods?

Reply:
No, I don't think so, the only time the element actually gets inserted is in the insert fn - the other things like rebalance and stuff don't